### ISSUE TITLE
[Feature Request]: Detect Page Navigation

### ISSUE THREAD

**Issue by rdelagato** [puppeteer/puppeteer#7819]
### Bug description

***What steps will reproduce the problem?***
I would like to detect if any action creates a page navigation event and if not terminate the tab.waitForPageNavigation. Example:
```
let element = await tab.$('div.search')
await Promise.all([
    element.click(),
    tab.waitForPageNavigation({waitUntil: 'networkIdle2', timeout: 15000})
])
```

***What is the expected result?***
Is there a variable that can be passed into waitForPageNavigation that can reject the promise if a navigation event has not been fired in X milliseconds. So I am clear, I am not referring to the timeout variable that waits up to Y seconds for the function to resolve. I would like both. I want the pageNavigation to wait X seconds for a navigation event to be fired and then reject if it has not been fired in that timeframe and then I would like for the timeout variable to wait up to Y seconds for that navigation event to be resolved.

```
let element = await tab.$('div.search')
await Promise.all([
    element.click(),
    tab.waitForPageNavigation({waitUntil: 'networkIdle2', pageNavigation: 2000, timeout: 15000})
])
```

***What happens instead?***
Currently, the code has to wait for the entire timeout in waitForPageNavigation even if a page navigation event has not been fired, which slows down testing. I don't want to alter the timeout for waitForPageNavigation because some pages take awhile to load so I would like to keep that time in milliseconds.

### Puppeteer version

10.4.0

### Node.js version

14.15.2

### npm version

6.14.9

### What operating system are you seeing the problem on?

Windows

**Comment by surlyhacker** [comment]
I had this same problem, but found that this approach works fine:
```
    var [response2] = await Promise.all([
      page.waitForRequest( (req) => {
        return req.isNavigationRequest();
      },{timeout: 10000}),
      page.waitForNavigation({waitUntil: "domcontentloaded", timeout: 0}),  
      page.click('#button'),  
      
    ]);
```

Here is the behavior I found in these scenarios with the above code:
1) If the `click()` does not result in a page navigation HTTP request being _initiated_ within the `waitForRequest()` `timeout`, it will throw a `TimeoutError`  (HTTP requests that would not trigger navigation are ignored)
2) If the `click()` does result in a navigation being initiated, it will timeout in accordance with the separate `waitForNavigation()` `timeout`

Note that Puppeteer's determination of `isNavigationRequest()` [uses this logic](https://github.com/puppeteer/puppeteer/blob/7748730163bc1a14cbb30881809ea529844f887e/src/common/HTTPRequest.ts#L173):
```
this._isNavigationRequest =
      event.requestId === event.loaderId && event.type === 'Document';
```

In other words, it corresponds to requests that are of the "[Document](https://chromedevtools.github.io/devtools-protocol/tot/Network/#type-ResourceType)" type as would be seen in the "type" column in the Chrome DevTools Network tab.

**Comment by surlyhacker** [comment]
Also if possible I'd suggest you change the title of this feature request, even if it gets closed.
"Detect Page Navigation" is already possible.  Maybe something like "wait for page navigation to be initiated (not completed)" would be more descriptive.  It will help with people doing Google searches for this code.
